### PR TITLE
PMM-7353 Create API Client for each DB cluster.

### PR DESCRIPTION
**Pull request by BupycHuk**
[PMM-7353](https://jira.percona.com/browse/PMM-7353)

- [ ] Build: https://github.com/Percona-Lab/pmm-submodules/pull/1677

- [ ] Links to other linked pull requests (optional).
https://github.com/Percona-Lab/pmm-submodules/pull/1677
https://github.com/percona/pmm-managed/pull/723
https://github.com/percona-platform/dbaas-controller/pull/173
https://github.com/percona-platform/dbaas-api/pull/121
---
- [ ] Tests passed.
- [ ] Feature build pass.
- [ ] (Re)requested review.
- [ ] Fix conflicts with target branch.
- [ ] Update API dependency.
- [ ] Update jira ticket description if needed.
- [ ] Attach screenshots/console output to confirm new behavior to jira ticket, if applicable.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -30,7 +30,7 @@ services:
 
     ports:
       - 127.0.0.1:80:80
-      - 127.0.0.1:443:443
```
Comment (palash25) [inline docker-compose.yml:33]: why this change?

File: services/grafana/client.go [hunk 2]
```diff
@@ -300,7 +303,34 @@ func (c *Client) testDeleteUser(ctx context.Context, userID int, authHeaders htt
 	return c.do(ctx, "DELETE", "/api/admin/users/"+strconv.Itoa(userID), "", authHeaders, nil, nil)
 }
 
-func (c *Client) testCreateAPIKey(ctx context.Context, name string, role role, authHeaders http.Header) (int, string, error) {
+func (c *Client) CreateAdminAPIKey(ctx context.Context, name string) (int, string, error) {
```
Comment (palash25) [inline services/grafana/client.go:306]: toplevel comment missing

File: docker-compose.yml (hunk 1 above)
Comment (BupycHuk) [reply docker-compose.yml:33]: To test monitoring of DB Services in DBaaS.

File: services/grafana/client.go [hunk 3]
```diff
@@ -300,7 +303,34 @@ func (c *Client) testDeleteUser(ctx context.Context, userID int, authHeaders htt
...
+	headers, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return 0, "", fmt.Errorf("cannot get headers from metadata %v", headers)
+	}
+	fmt.Println(headers)
```
Comment (palash25) [inline services/grafana/client.go:311]: please remove this

File: services/grafana/client.go [hunk 4]
```diff
@@ -300,7 +303,34 @@ func (c *Client) testDeleteUser(ctx context.Context, userID int, authHeaders htt
...
+	// get authorization from headers.
+	authorizationHeaders := headers.Get("Authorization")
+	fmt.Println(authorizationHeaders)
+	cookieHeaders := headers.Get("grpcgateway-cookie")
+	fmt.Println(cookieHeaders)
```
Comment (palash25) [inline services/grafana/client.go:316]: i am guessing all these `fmt.Println` were for debugging and you forgot to remove them, if yes then please remove them otherwise maybe log it instead if we want to see this information

Comment (BupycHuk) [reply services/grafana/client.go:316]: yes, correct

File: services/grafana/client.go (hunk 2 above)
Comment (palash25) [inline services/grafana/client.go:306]: // CreateAdminAPIKey creates API key to be used by DBaaS clusters

something like that

Comment (BupycHuk) [reply services/grafana/client.go:307]: Done

**Review by jprukner — APPROVED** [review]

**Review by palash25 — APPROVED** [review]

File: services/grafana/client.go [hunk 5]
```diff
@@ -300,7 +303,32 @@ func (c *Client) testDeleteUser(ctx context.Context, userID int, authHeaders htt
...
 
-func (c *Client) testCreateAPIKey(ctx context.Context, name string, role role, authHeaders http.Header) (int, string, error) {
+// CreateAdminAPIKey creates API key with Admin role and provided name.
+func (c *Client) CreateAdminAPIKey(ctx context.Context, name string) (int, string, error) {
+	headers, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return 0, "", fmt.Errorf("cannot get headers from metadata %v", headers)
+	}
+	// get authorization from headers.
+	authorizationHeaders := headers.Get("Authorization")
```
Comment (oter) [inline services/grafana/client.go:313]: maybe create a constant values for `Authorization`? Since, I see, it is used in many places.

File: services/grafana/client.go [hunk 6]
```diff
@@ -300,7 +303,32 @@ func (c *Client) testDeleteUser(ctx context.Context, userID int, authHeaders htt
...
+	cookieHeaders := headers.Get("grpcgateway-cookie")
+	if len(authorizationHeaders) == 0 && len(cookieHeaders) == 0 {
```
Comment (oter) [inline services/grafana/client.go:315]: what if both are set?

**Review by oter — APPROVED** [review]

**Review by jprukner — APPROVED** [review]

File: services/grafana/client.go (hunk 6 above)
Comment (BupycHuk) [reply services/grafana/client.go:354]: then both will be added to headers, no problem :)

File: services/grafana/client.go [hunk 7]
```diff
@@ -300,36 +311,89 @@ func (c *Client) testDeleteUser(ctx context.Context, userID int, authHeaders htt
...
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) authHeadersFromContext(ctx context.Context) (http.Header, error) {
+	headers, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, fmt.Errorf("cannot get headers from metadata %v", headers)
```
Comment (oter) [inline services/grafana/client.go:349]: `headers` will be nil because `ok == false`

File: services/management/dbaas/psmdb_cluster_service.go [hunk 8]
```diff
@@ -231,6 +247,9 @@ func (s PSMDBClusterService) DeletePSMDBCluster(ctx context.Context, req *dbaasv
 		return nil, err
 	}
 
+	// ignore if API Key is not deleted.
+	_ = s.grafanaClient.DeleteAPIKeysWithPrefix(ctx, fmt.Sprintf("%s-%s", req.KubernetesClusterName, req.Name))
```
Comment (oter) [inline services/management/dbaas/psmdb_cluster_service.go:251]: maybe at least log it? since we will never know if it works or not.

File: services/management/dbaas/xtradb_cluster_service.go [hunk 9]
```diff
@@ -311,6 +327,9 @@ func (s XtraDBClusterService) DeleteXtraDBCluster(ctx context.Context, req *dbaa
 		return nil, err
 	}
 
+	// ignore if API Key is not deleted.
+	_ = s.grafanaClient.DeleteAPIKeysWithPrefix(ctx, fmt.Sprintf("%s-%s", req.KubernetesClusterName, req.Name))
```
Comment (oter) [inline services/management/dbaas/xtradb_cluster_service.go:331]: same here

File: services/grafana/client.go (hunk 7 above)
Comment (BupycHuk) [reply services/grafana/client.go:349]: No, it won't be nil :) It will be zero value.
I'll update error message

**Review by palash25 — APPROVED** [review]

**Review by jprukner — APPROVED** [review]

**Review by palash25 — APPROVED** [review]
